design/include-cleaner.md

```diff
@@ -0,0 +1,139 @@
+# Include Cleaner
+
+The common pattern in C and C++ programs is that the source files often contain
+unnecessary includes. Most of the time, these includes were added at some point
+and actually used for some time, but over time the program evolved and the
+symbols defined in these headers were no longer used. Keeping an eye on what
+headers are actually required to build the file and manually updating them is
+time-consuming and error-prone. Over time, the stale inclusions can pile up and
+increase build times and complicate the dependncy graph.
+
+Include Cleaner is a
+[IWYU](https://github.com/include-what-you-use/include-what-you-use)-inspired
+feature in clangd that automates the process of maintaining a necessary set of
+included headers. It now is available in "preview" mode with an incomplete set
+of capabilities. If you experience any bugs, please submit a bug report in
+[clangd/issues](https://github.com/clangd/clangd/issues).
+
+{:.v14}
+
+## How it works
+
+Include Cleaner issues diagnostics fro includes that are present but not used
+in the main file. When you open a file, clangd will analyze the symbols the
+file uses and mark all headers defining these symbools as "used". The warnings
+will be issues for the headers that are included but not marked as "used".
+Example:
+
+```c++
+// foo.h
+class Foo {};
+```
+
+```c++
+// bar.h
+class Bar {};
+```
+
+```c++
+// main.cpp
+
+#include "foo.h"
+#include "bar.h" // <- Will be marked as unused and suggested to be removed.
+
+int main() {
+  Foo f;
+}
+```
+
+Here, `main.cpp` only makes use of symbols from `foo.h` and removing `#include
+"bar.h"` prevents unnecessary parsing of `bar.h` and allows breaking the
+dependncy on it.
+
+### Conservative filtering
+
+Include Cleaner strives to provide safe diagnostics to reduce the posibility of
+breaking the build. The tool will consider that every symbol in the "expansion"
+of the main file symbols to count as a usage. Example:
+
+```c++
+// foo.h
+
+int foo() { return 42; }
+```
+
+```c++
+// macro.h
+
+#include "foo.h"
+
+#define FOO foo
+```
+
+```c++
+// main.cpp
+
+#include "macro.h"
+
+int main() {
+  FOO();
+}
+```
+
+In this case, both `macro.h` and `foo.h` will be marked as "used" because the
+macro expansion contains the function call. `auto` will have a similar effect.
+
+### IWYU pragmas
+
+IWYU tool offers a set of
+[pragmas](https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/IWYUPragmas.md),
+the most notable of which are:
+
+- `IWYU pragma: keep` indicates the inclusion should not be removed from the
+  main file
+- `IWYU pragma: private` indicates that the current header is an implementation
+  detail and another one should be included instead
+
+Right now, only the `keep` pragma is implemented, `private` will be implemented
```

Comment:
Maybe just claim we support both here? as we're about to do so and this is design, not implementation so fine to talk about final state.